### PR TITLE
Tracer can output a compact tracing log file

### PR DESCRIPTION
**Pull request by Dr15Jones**
#### PR description:

Added an option to the Tracer service to output a compact log file of the transitions seen in the framework job. 
Included a python program to convert the file into more humanly readable output.

#### PR validation:

Code compiles. Visual inspection of a file generated from a simple configuration shows the expected values.

### CODE REVIEW

**Comment by Dr15Jones** [comment]
please test

**Comment by Dr15Jones** [comment]
hold

considering putting this ability elsewhere

**Comment by Dr15Jones** [comment]
code-checks

**Comment by Dr15Jones** [comment]
please test

**Comment by Dr15Jones** [comment]
untold

File: FWCore/Services/bin/edmTracerCompactLogViewer.py [hunk 1]
```diff
@@ -0,0 +1,577 @@
...
+  Once the graph is completed, the program outputs the list of modules
+  which had the greatest total stall times. The list is sorted by
+  total stall time and written in descending order. In addition, the
+  list of all stall times for the module is given.
+
+  There is an inferior alternative (an old obsolete way).
+  Instead of using the StallMonitor Service, you can use the
+  Tracer Service.  Make sure to use the 'printTimestamps' option
+  cms.Service("Tracer", printTimestamps = cms.untracked.bool(True))
+  There are problems associated with this and it is not recommended.'''
```
Comment (makortel) [inline FWCore/Services/bin/edmTracerCompactLogViewer.py:49]: How accurate is this help message? By quick look it looks mostly being copied from `edmStreamStallGrapher.py`.

File: FWCore/Services/bin/edmTracerCompactLogViewer.py [hunk 2]
```diff
@@ -0,0 +1,577 @@
...
+#  postModuleTransition = 'm',
+#  preESModulePrefetching = 'Q',
+#  postESModulePrefetching = 'q',
+#  preESModule = 'N',
+#  postESModule = 'n',
+#  preESModuleAcquire = 'B',
+#  postESModuleAcquire = 'b',
+#  preFrameworkTransition = 'F',
+#  postFrameworkTransition = 'f'
+#};
```
Comment (makortel) [inline FWCore/Services/bin/edmTracerCompactLogViewer.py:73]: Is this commented out piece useful here?

File: FWCore/Services/bin/edmTracerCompactLogViewer.py [hunk 3]
```diff
@@ -0,0 +1,577 @@
...
+
+
+#Special names
+kSourceFindEvent = "sourceFindEvent"
+kSourceDelayedRead ="sourceDelayedRead"
+
+class Phase (object):
```
Comment (makortel) [inline FWCore/Services/bin/edmTracerCompactLogViewer.py:80]: Maybe add a comment that the values must correspond to those of `enum class Phase` in `FWCore/Services/plugins/tracer_setupFile.cc`?

File: FWCore/Services/plugins/tracer_setupFile.cc [hunk 4]
```diff
@@ -0,0 +1,814 @@
...
+#include "FWCore/Utilities/interface/TimingServiceBase.h"
+
+using namespace edm::service::monitor_file_utilities;
+
+namespace {
+  using duration_t = std::chrono::microseconds;
+  using clock_t = std::chrono::steady_clock;
+  auto now = clock_t::now;
+
+  enum class step : char {
```
Comment (makortel) [inline FWCore/Services/plugins/tracer_setupFile.cc:39]: Could this enum start with upper case letter, i.e. `Step`?

File: FWCore/Services/plugins/tracer_setupFile.cc [hunk 5]
```diff
@@ -0,0 +1,814 @@
...
+#include "FWCore/Framework/interface/ESRecordsToProductResolverIndices.h"
+
+#include "FWCore/Utilities/interface/TimingServiceBase.h"
+
+using namespace edm::service::monitor_file_utilities;
+
+namespace {
+  using duration_t = std::chrono::microseconds;
+  using clock_t = std::chrono::steady_clock;
+  auto now = clock_t::now;
```
Comment (makortel) [inline FWCore/Services/plugins/tracer_setupFile.cc:37]: Can this global variable be `const`? Or use a function for the shorthand?

File: FWCore/Services/bin/edmTracerCompactLogViewer.py (hunk 2 above)
Comment (makortel) [reply FWCore/Services/bin/edmTracerCompactLogViewer.py:73]: Would it be useful define these as "enum" similarly to the `Phase` class?

File: FWCore/Services/bin/edmTracerCompactLogViewer.py (hunk 1 above)
Comment (Dr15Jones) [reply FWCore/Services/bin/edmTracerCompactLogViewer.py]: fixed

File: FWCore/Services/plugins/tracer_setupFile.cc (hunk 4 above)
Comment (Dr15Jones) [reply FWCore/Services/plugins/tracer_setupFile.cc]: done

File: FWCore/Services/plugins/tracer_setupFile.cc (hunk 5 above)
Comment (Dr15Jones) [reply FWCore/Services/plugins/tracer_setupFile.cc]: done

File: FWCore/Services/bin/edmTracerCompactLogViewer.py (hunk 3 above)
Comment (Dr15Jones) [reply FWCore/Services/bin/edmTracerCompactLogViewer.py:52]: done

File: FWCore/Services/bin/edmTracerCompactLogViewer.py (hunk 2 above)
Comment (Dr15Jones) [reply FWCore/Services/bin/edmTracerCompactLogViewer.py:44]: I hadn't done that as the values are used in one and only one spot in the entire code.

**Comment by Dr15Jones** [comment]
please test

**Comment by makortel** [comment]
The unit test fails in IBs

**Comment by makortel** [comment]
+core

**Comment by Dr15Jones** [comment]
please test

**Comment by Dr15Jones** [comment]
NOTE: It is impossible for this PR to affect any test results as the change can only be seen if one uses a brand new parameter in the Tracer service, of which no code in CMSSW uses.

**Comment by antoniovilela** [comment]
+1

**Comment by makortel** [comment]
Comparison differences are related to https://github.com/cms-sw/cmssw/issues/39803

**Pull request merged**